crates/plex-api/src/server/library.rs

```diff
@@ -125,6 +127,113 @@ where
     }
 }
 
+/// A single media format for a `MediaItem`.
+pub struct Media<'a> {
+    client: &'a HttpClient,
+    media: &'a MediaMetadata,
+}
+
+impl<'a> Media<'a> {
+    /// The different parts that make up this media. They should be played in
+    /// order.
+    pub fn parts(&self) -> Vec<Part> {
+        self.media
+            .parts
+            .iter()
+            .map(|part| Part {
+                client: self.client,
+                part,
+            })
+            .collect()
+    }
+
+    /// The internal metadata for the media.
+    pub fn metadata(&self) -> &MediaMetadata {
+        self.media
+    }
+}
+
+/// One part of a `Media`.
+pub struct Part<'a> {
+    client: &'a HttpClient,
+    part: &'a PartMetadata,
+}
+
+impl<'a> Part<'a> {
+    /// The length of this file on disk in bytes.
+    pub fn len(&self) -> Option<u64> {
+        self.part.size
+    }
+
+    /// Downloads the original media file for this part. Optionally a range of bytes within the
+    /// file can be requested allowing for resumable transfers.
+    pub async fn download<R: RangeBounds<u64>>(&self, range: R) -> Result<AsyncBody> {
```

Comment:
I have mixed feelings about returning AsyncBody in the `download()` method. What do you think about either slightly renaming the method (to `request` maybe?) or rewriting the method so that it will accept an object with the `Write` trait implemented and write the result into it?

Reply:
I think passing a Write trait makes a lot of sense, then we can avoid exposing the specific Http implementation used.

Reply:
`Write` or `AsyncWrite` do you think?

Reply:
I think we need to go with `Write` for now, as it will be easier for us and the end-users. We'll be able to switch to `AsyncWrite` later with a breaking change if it is worth it.

Reply:
... but I didn't know that `AsyncWrite` existed until 15 minutes ago. So take my thoughts with a grain of salt.

Reply:
So looking more at this, I understand the rationale of not relying on the non-standard [`AsyncWrite`](https://docs.rs/futures/latest/futures/io/trait.AsyncWrite.html) but implementing it is trivial, just a one or two line change to the patch. Supporting `Write` means either switching to a sync download in this case, which seems bad for large files, or doing a bunch of work to poll the async download and then wait for the blocking `Write` to accept those bytes. If you want me to do that then I will, but I'd argue that, assuming we want this to be async in the future anyway, once `AsyncWrite` or something like it gets standardized then we'll have to make an API change whatever and switching callers from one kind of async writer to another is likely to be easier than switching from a sync writer to an async writer. Let me know what you prefer to go with.

Reply:
If having an `AsyncWrite` doesn't really complicate the implementation and usage, then sure, go ahead.